### ISSUE TITLE
/nix/store strings are unconditionally stripped from kernelParams

### ISSUE THREAD

**Issue by blitz** [cyberus-technology/cbspkgs-public#26]
# Overview

I use `linuxBootItem` from `lib/sotest.nix` to generate a boot item for a NixOS image. The command line of Linux is:

```
init=/nix/store/8c7fhsa0yk9v7frj27pn89qmswy0fyfr-nixos-system-nixos-21.05pre-git/init initrd=initrd
```

The `/nix/store` path does not refer to anything that needs to be in the test bundle. It refers to the init binary in the initramfs that I ship. Nevertheless the Linux command line gets modified and the /nix/store prefix for the init binary is lost.

# Current Behavior

Any `/nix/store` string is stripped from `kernelParams`.

# Expected Behavior

`kernelParams` is not modified.

**Comment by parthy** [comment]
This is interesting. There might be edge cases where you want to refer to contents in the binary package (meaning the prefix would need to be stripped) vs. specifying raw, unrelated values containing the prefix kind of "by accident". I would propose something where the caller can explicitly mark sections that should be used 1:1, bypassing the substitution.

**Comment by tfc** [comment]
> This is interesting. There might be edge cases where you want to refer to contents in the binary package (meaning the prefix would need to be stripped) vs. specifying raw, unrelated values containing the prefix kind of "by accident". I would propose something where the caller can explicitly mark sections that should be used 1:1, bypassing the substitution.

If we had a rule that states that only the first word of every boot/load line is substituted: then you would still have the freedom to provide a nix store path in the cmd line, or if you want to have the `/nix/store` stripped, you can just do that and then effectively have the choice between both, right? Would that meet your idea/requirement or did i miss something?

**Comment by parthy** [comment]
Having both is what we want in the end. The use cases I have in mind are:

```nix
# Referencing a "local" nix store path, which needs to be adjusted
{
  exec = "${bzImage}/bzImage initrd=${initrd}/initrd";
  extra_files = [
    "${initrd}/initrd"
  ];
}

# Referencing a nix store path in a different context (in this case, the initrd)
{
  exec = "${bzImage}/bzImage init=/nix/store/.../init-script-inside-initrd";
  load = [
    "${initrd}/initrd"
  ];
}
```

Logically, the indicator for path substitution is not a positional one, but a logical one. Whenever a nix store path in a project config is referring to a path within the bundle zip file, it needs to be adjusted in the same way as the store path is adjusted when it is included in the zip file in the first place. This is where your suggestion is a bit unfortunate, because we would hardcode the logic behind the archive creation in the library usage context. The specific method of just stripping the `/nix/store` prefix is an implementation detail and could just as well be changed to something like `/nix/store/ -> nixStore/` and the user of the library should not need to know that, nor should it affect him if that detail is changed under the hood.

From a conceptual perspective, the library should compensate for whatever changes it makes to the paths and leave all unrelated paths alone. This can be tricky when paths are present in the archive, but also in an unrelated context (e.g., a path is present in the initrd and that is what the cmdline argument refers to, but the same path is also loaded as a module somewhere in the project config). Hoping that this is more of an unlikely edge case, we could define a rule that in that case, the substitution will always take place. Although an explicit marker indicating that a given path should not be adjusted by the library would work either way.

**Comment by parthy** [comment]
I can fill up the command line with whatever I want and parse that in userspace via `/proc/cmdline`. That might not be a specified mechanism from the purist Linux point of view, but it's possible. And if I put any file path on the command line that is then invalidated by the library, the path on the command line needs to be adjusted for that. Furthermore, this mechanism is not limited to Linux. I can specify any kernel, any command line, and any extra file I want. My kernel can parse the command line and needs to be able to make sense of it.

> it's easy to add whatever you need

How? As a user of the library, I have no way of knowing where the files I referenced via nix attributes end up. Nor should I have to. I maintain my strong position that the library should transparently compensate for what it does behind the scenes. Anything that's valid with the original store paths should remain valid after the substitution.

**Comment by kaldonir** [comment]
@blitz This should be fixed with #34, right?

**Comment by blitz** [comment]
Yes 🙏
